pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -23,431 +21,159 @@ const (
-
-  // Build dependency trees for all image streams in the current namespace
-  $ openshift ex build-chain
-
-  // Build dependency trees for all image streams across all namespaces
-  $ openshift ex build-chain --all`
+  // Build the dependency tree across all namespaces for the specified image stream tag found in 'test' namespace
+  $ %[1]s build-chain centos7 -n test --all`
 )
 
-// Node is a representation of an image stream inside a tree
-type Node struct {
-	FullName string   `json:"fullname"`
-	Tags     []string `json:"tags,omitempty"`
-	Edges    []*Edge  `json:"edges,omitempty"`
-	Children []*Node  `json:"children,omitempty"`
-}
-
-// String helps in dumping a tree in AST format
-func (root *Node) String() string {
-	tree := ""
-	tree += root.FullName
+// BuildChainOptions contains all the options needed for build-chain
+type BuildChainOptions struct {
+	name       string
+	defaultNs  string
+	namespaces []string
+	tag        string
+	output     string
 
-	for _, n := range root.Children {
-		tree += "(" + n.String() + ")"
-	}
+	allNamespaces bool
 
-	return tree
-}
-
-// Edge represents a build configuration relationship
-// between two nodes
-//
-// Note that this type has no relation with the dot.Edge
-// type
-type Edge struct {
-	FullName string `json:"fullname"`
-	To       string `json:"to"`
-}
-
-// NewEdge adds a new edge on a parent node
-//
-// Note that this function has no relation
-// with the dot.Edge type
-func NewEdge(fullname, to string) *Edge {
-	return &Edge{
-		FullName: fullname,
-		To:       to,
-	}
+	oc client.Interface
 }
 
 // NewCmdBuildChain implements the OpenShift experimental build-chain command
-func NewCmdBuildChain(f *clientcmd.Factory, parentName, name string) *cobra.Command {
+func NewCmdBuildChain(name, fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	options := &BuildChainOptions{}
 	cmd := &cobra.Command{
-		Use:     fmt.Sprintf("%s [IMAGESTREAM:TAG | --all]", name),
-		Short:   "Output build dependencies of a specific image stream",
+		Use:     "build-chain [IMAGESTREAM:TAG] [OPTIONS]",
+		Short:   "Output the inputs and dependencies of your builds",
 		Long:    buildChainLong,
-		Example: buildChainExample,
+		Example: fmt.Sprintf(buildChainExample, fullName),
 		Run: func(cmd *cobra.Command, args []string) {
-			err := RunBuildChain(f, cmd, args)
+			err := options.Complete(f, cmd, args, out)
+			cmdutil.CheckErr(err)
+
+			err = options.Validate(cmd, args)
+			cmdutil.CheckErr(err)
+
+			err = options.RunBuildChain(f, cmd, args)
 			cmdutil.CheckErr(err)
 		},
 	}
 
-	cmd.Flags().Bool("all", false, "Build dependency trees for all image streams")
-	cmd.Flags().Bool("all-tags", false, "Build dependency trees for all tags of a specific image stream")
-	cmd.Flags().StringP("output", "o", "json", "Output format of dependency tree(s)")
+	cmd.Flags().BoolVar(&options.allNamespaces, "all", false, "Build dependency tree for the specified image stream tag across all namespaces")
```

Comment:
> how can we separate those two here? check the fullName and if its not oadm or openshift admin then it's a user?

You can make an issue so we can talk about it.  I don't want to hold up this pull for it.  The current impl works correctly for cluster-admins, it's just weird.